torch/optim/optimizer.py

```diff
@@ -114,17 +114,24 @@ def _cuda_graph_capture_health_check(self):
                       "instance, capturable=True can impair performance, and you should set capturable=False.")
                 self._warned_capturable_if_run_uncaptured = True
 
+    def hook_code(self):
```

Comment:
Adding hooks to optimizer sounds ok to me but we most likely want to make this a proper hook system like we have for Module: https://pytorch.org/docs/stable/generated/torch.jit.ScriptModule.html?highlight=register_#torch.jit.ScriptModule.register_forward_hook

Reply:
We already have special code to instrument optimizer; this is more or less an extension of that code. Whereas hooks would be a public API change.

Reply:
@albanD do you have any thoughts on this with more context?

Reply:
The thing is that this is not going to compose well with whoever else tries to hook into the optimizer, change the way they behave or override the default step() function.

To be honest, the current hack done by the profiler is not very satisfying either as swapping functions around needs to be done in a very strict order to ensure you get the behavior you are looking for at the end.

If we have a proper step hook on the optimizer, you can actually completely remove the current hack.

Reply:
I see. If Im understanding correctly, you are thinking the entire `_hook_for_profile` was a hack in the first place and we should have gone with proper step hook so we are not building more stuff on top of this hack?

how much work would it be to have a proper hook API for both internal&external? for the sake of simplicity, we can consider "internal" hook with plan to have it public as well?

Reply:
I agree, the current technique is a hack. However I don't think this is *more* of a hack than the current approach. I have no issue with a proper hook API, however I expect that would take a while to design and go through RFC. (And it's not clear who such a thing would compose with what distributed is proposing for optimizers.) So my inclination is to go with this, but switch to hooks if they are added. WDYT?

Reply:
my only concern with sticking with the approach in the PR is if this extra hack makes the migration to the proper hook much harder, which doesn't seem like it.

I'd suggest that we leave a comment here saying we were working around the lack of proper optimizer hook and we are committed to migrating over if we ever cross that bridge.